docs/proposals/20200220-cluster-resource-set.md

```diff
@@ -253,10 +262,86 @@ When a `ClusterResourceSet` is deleted, it will be removed from the `bindings` l
 
-As a note, if providing different values for some fields in the resources for different clusters is needed such as CNIs podCIDRs, some templating mechanism for variable substitution in the resources can be used, but not provided by `ClusterResourceSet`.
+#### `Reconcile` mode
+
+***Detecting changes***
+`ClusterResourceBindings` will contain consistent hash for the resource/s definitions. We will use this to detect changes, comparing the hash of the current resource/s definition with the one stored in the `ClusterResourceBindings`.
+
+Note that this hash will change when any of the resources is updated, a resource is added or a resource is removed. This means that all resources in the same `ConfigMap` or `Secret`, and not only the one that changed, will be reapplied in any of these 3 cases. It also means that resources removed from `ConfigMap` or `Secret` won't be deleted from the target clusters.
+
+In the next *before-after* example we can see that only one resource has changed (`ConfigMap` `calico-configmap`). However, all the 3 resources (`calico-secret1`, `calico-secret2` and `calico-configmap`) will be reapplied.
+
+Before:
+```yaml
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: calico-addon
+data:
+  calico1.yaml: |-
+     kind: Secret
+     apiVersion: v1
+     metadata:
+      name: calico-secret1
+      namespace: mysecrets
+      ---
+     kind: Secret
+     apiVersion: v1
+     metadata:
+      name: calico-secret2
+      namespace: mysecrets
+  calico2.yaml: |-
+     kind: ConfigMap
+     apiVersion: v1
+     metadata:
+      name: calico-configmap
+      namespace: myconfigmaps
+     data:
+       key: "original value"
+```
+
+After:
+```yaml
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: calico-addon
+data:
+  calico1.yaml: |-
+     kind: Secret
+     apiVersion: v1
+     metadata:
+      name: calico-secret1
+      namespace: mysecrets
+      ---
+     kind: Secret
+     apiVersion: v1
+     metadata:
+      name: calico-secret2
+      namespace: mysecrets
+  calico2.yaml: |-
+     kind: ConfigMap
+     apiVersion: v1
+     metadata:
+      name: calico-configmap
+      namespace: myconfigmaps
+     data:
+       key: "value that changed"
+```
+***Drift***
+
+The proposed solution only deals with changes in the resources' definitions and not with changes in the real objects in the workload clusters. If those objects are modified or deleted in the workload clusters, the `ClusterResourceSet`'s controller won't do anything and they will remain unchanged until their definition in the management cluster is updated.
+
+This could potentially be mitigated by:
+ * Implementing a "periodic" reconciliation mode where resources are reapplied with a certain frequency even their hash hasn't changed.
```

Comment:
Periodic reconciliation has proven to be a bit of a headache for us. This tended to delete references to secrets that had to be recreated while leaving the original one behind causing a huge list of secrets in a cluster in a few hours. If we agree to this approach I would recommend approaching it with care.

Reply:
+1 to track these concerns after the proposed mitigation, and possibly expand on the specific problem.
Also, isn't this a risk that we have when we re-apply (assuming the behaviour is the same)

Reply:
@alejandroEsc would you want to add these concerns to the PR yourself? I think you should be able to push to this branch, if not, let me know!

Or if you prefer we can take it offline and I'll add it.

Reply:
Isn't this also an issue every time something is changed in the ConfigMap/Secret in the management cluster?

Reply:
I wonder in which case that issue occurs. It feels like a fundamental limitation of the way we would reapply the resources? (depending on how exactly we reapply I would expect that other controllers can modify the resources without us overwriting the changes)

Reply:
@sbueringer good point. I see what you are saying but I'm not completely sure what kind of "deletions" @alejandroEsc was referring to and how those happened.

If we assume the problem is that the CRS controller was overwriting (or more than overwriting, nullfyng) fields set by other controllers/entities, then yeah, the same problem would be present with the proposed solution (the problem in this scenario is not the periodicity, it's the apply mechanism, the periodicity just makes it worse). However, I don't see why we couldn't update/patch the objects in a way that we only modify whichever fields are specified in the CRS definitions, that's a common thing in controllers. We could even use server-side apply but I'm not even sure it's necessary.

@alejandroEsc could you provide a bit more details?

Reply:
@g-gaston I agree, sounds good. I think it should work by just patching the objects. SSA could be ideal, but unfortunately we are supporting workload clusters starting with 1.18 so SSA probably can't be used for those old clusters.

Reply:
To provide a bit more context, and totally my fault for dropping the ball here, but what would happen is that service accounts would be recreated and new tokens would be created. The previous tokens would remain in place, so we never saw a failure in the services due to disruptions but only a huge number of tokens in any one of the affected namespaces. This is what I mean by my comments above.

Reply:
Thx for the additional context!

@g-gaston Slight update regarding
> SSA could be ideal, but unfortunately we are supporting workload clusters starting with 1.18 so SSA probably can't be used for those old clusters.

We were talking about using SSA for workload clusters in the context of syncing labels to nodes. We are currently only exploring using SSA there. It could be that we use SSA and then only support the node label sync feature for workload clusters with Kubernetes >= v1.22 (no decisions are made).

We have to discuss it, but this could be an option for CRS as well.

(cc @fabriziopandini @enxebre)

Reply:
@sbueringer ahh good to know!

I would lean towards making this doc agnostic from the apply implementation for now. I would also vote to try to implement this without SSA (given the 1.18 constrain) and if we then realize there are other problems and/or we find other potential benefits of using SSA, we can come back to this doc and add it.

Reply:
The discussion about SSA is interesting, but I will defer this discussion to next iteration given that this part is not in the scope of the changes introduced by this PR (it is just an idea about risk mitigation). 

If it could help to unblock, let's be explicit about this:

```
This could be eventually mitigated by:
 * Implementing a "periodic" reconciliation mode where resources are reapplied with a certain frequency even their hash hasn't changed, but this is out of the scope of this proposal.
 ```

Reply:
I'm totally fine to merge the proposal as is. Maybe it's something we want to explore during implementation.

I think it would be just a serious limitation if the CRS would produce an ever increasing amount of secrets as soon as a ServiceAccount is deployed with CRS.

But imho defintely nothing that we have to solve in the proposal itself (as it's kind of an implementation detail to figure out a way it doesn't happen)